pkg/parser/translate.go

```diff
@@ -419,21 +418,52 @@ func fromUDPIngressV1Alpha1(log logrus.FieldLogger, ingressList []*v1alpha1.UDPI
 			"udpingress_name":      ingress.Name,
 		})
 
-		result.ServiceNameToServices[ingress.Namespace+"."+ingress.Name] = kongstate.Service{
-			Service: kong.Service{
-				Name:     kong.String(ingress.Namespace + "." + ingress.Name),
-				Protocol: kong.String("udp"),
-				Host:     kong.String(ingressSpec.Host),
-				Port:     kong.Int(ingress.Spec.TargetPort),
-			},
-			Routes: []kongstate.Route{
-				{
-					Route: kong.Route{
-						Protocols:    []*string{kong.String("udp")},
-						Destinations: []*kong.CIDRPort{{Port: kong.Int(ingressSpec.ListenPort)}},
-					},
+		for i, rule := range ingressSpec.Rules {
+			// validate the ports and servicenames for the rule
+			if !util.IsValidPort(rule.Port) {
+				log.Errorf("invalid UDPIngress: invalid port: %d", rule.Port)
+				continue
+			}
+			if rule.Backend.ServiceName == "" {
+				log.Errorf("invalid UDPIngress: empty serviceName")
+				continue
+			}
+			if !util.IsValidPort(rule.Backend.ServicePort) {
+				log.Errorf("invalid UDPIngress: invalid servicePort: %d", rule.Backend.ServicePort)
+				continue
+			}
+
+			// generate the kong Route based on the listen port
+			route := kongstate.Route{
+				Ingress: util.FromK8sObject(ingress),
+				Route: kong.Route{
+					Name:         kong.String(ingress.Namespace + "." + ingress.Name + "." + strconv.Itoa(i) + ".udp"),
+					Protocols:    kong.StringSlice("udp"),
+					Destinations: []*kong.CIDRPort{{Port: kong.Int(rule.Port)}},
 				},
-			},
+			}
+
+			// generate the kong Service backend for the UDPIngress rules
+			host := fmt.Sprintf("%s.%s.%d.svc", rule.Backend.ServiceName, ingress.Namespace, rule.Backend.ServicePort)
+			serviceName := fmt.Sprintf("%s.%s.%d.udp", ingress.Namespace, rule.Backend.ServiceName, rule.Backend.ServicePort)
+			service, ok := result.ServiceNameToServices[serviceName]
+			if !ok {
+				service = kongstate.Service{
+					Namespace: ingress.Namespace,
+					Service: kong.Service{
+						Name:     kong.String(serviceName),
+						Protocol: kong.String("udp"),
+						Host:     kong.String(host),
+						Port:     kong.Int(rule.Backend.ServicePort),
+					},
+					Backend: kongstate.ServiceBackend{
+						Name: rule.Backend.ServiceName,
+						Port: kongstate.PortDef{Mode: kongstate.PortModeByNumber, Number: int32(rule.Backend.ServicePort)},
+					},
+				}
+			}
+			service.Routes = append(service.Routes, route)
+			result.ServiceNameToServices[serviceName] = service
```

Comment:
I think only need do line 466 assignment when !OK. 
service is a map, line 465 changes automatically reflected within result if service already existing.

Reply:
Similarly to how this is done in TCP, this needs to be done in any case:

the `!ok` check will simply generate a new `konstate.Service` if one didn't exist in the map.

Right? :thinking:

Reply:
@ccfishk and I talked about this in slack today, we're both OK with holding on this and other smaller optimizations for now given the desire to get this released soon.